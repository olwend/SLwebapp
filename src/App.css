.main-container {
  flex: 1;

  overflow: auto;
}

.sticky-nav-bottom {
  position: sticky;
  bottom: 0;
  background-color: #f8f9fa
}