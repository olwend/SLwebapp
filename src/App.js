import React, { useState, useEffect } from "react";
import { withRouter } from "react-router-dom";
import { Container } from "reactstrap";
import Routes from "./routes/Routes";
import "./App.css";
import { Auth } from "aws-amplify";
import { useDispatch, useSelector } from "react-redux";
import { setIsLabRunning, getSession, setRemainingMins } from './redux/slices/sessionSlice';
import { setIsLoaded } from './redux/slices/userSlice';
import { setUserData } from "./utils/authentication";
import { loadSession } from "./utils/session";

function App() {

  // Set initial states
  const [isAuthenticating, setIsAuthenticating] = useState(true);
  const dispatch = useDispatch();
  const sessionInfo = useSelector(getSession);
  const minInMs = 60000;
  const [ wasCalled, setWasCalled ] = useState(false);
  const [ firstCall, setFirstCall ] = useState(true);

  useEffect(() => {
    onLoad();
  // eslint-disable-next-line
  },[]);

  // This is the code that gets the timer
  // eslint-disable-next-line
  useEffect(() => {
    if(sessionInfo.isLabActive && !wasCalled){
      let timerVal = sessionInfo.session.timer.value * 60 * 1000;
      if(firstCall){
        setFirstCall(false);
        remainingTime(timerVal);
      }
      setWasCalled(true);
      let timerSec = sessionInfo.remainingMins === 0 ? 1000 : minInMs;
      
      setTimeout(() => {
        remainingTime(timerVal);
      }, timerSec);
    }
  });

  async function onLoad() {

  // Check if the user is authenticated
    try {
      await Auth.currentAuthenticatedUser({ bypassCache: false })
        .then((user) => { 
          setUserData(dispatch, user);
          loadSession(dispatch, user.username);
          dispatch(setIsLabRunning(false));
        })
        .then(setIsAuthenticating(false))
        .catch((err) => {
          console.log(err)
          dispatch(setIsLoaded(true))
        });      
    } catch (e) {
      if (e !== "No current user") {
        alert(e);
      }
    }
  }

  function remainingTime(timer) {
      const endTime = sessionInfo.session.startDateTime + timer;
      const timeNow = Date.now();
      let mins = 0;
      
      if(endTime > timeNow){
        mins = ((endTime-timeNow) / minInMs).toFixed(0);
        dispatch(setIsLabRunning(true));
      }

      dispatch(setRemainingMins(mins));
      setWasCalled(false);
    // }
    }
    
  return (
    !isAuthenticating && (
      <Container fluid className="px-0 vh-100">
        <Routes />
      </Container>
    )
  );
}

export default withRouter(App);
