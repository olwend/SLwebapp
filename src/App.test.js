import React from 'react'

import { shallow } from 'enzyme'

import { BrowserRouter as Router } from 'react-router-dom'

import App from './App'

test('Renders properly', () => expect(shallow(<Router><App /></Router>).dive()).toMatchSnapshot())
