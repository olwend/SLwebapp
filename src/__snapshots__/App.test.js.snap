// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders properly 1`] = `
<ContextProvider
  value={
    Object {
      "history": Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      },
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "match": Object {
        "isExact": true,
        "params": Object {},
        "path": "/",
        "url": "/",
      },
      "staticContext": undefined,
    }
  }
>
  <ContextProvider
    value={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(App) />
  </ContextProvider>
</ContextProvider>
`;
